File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/ScaleShapeMouseCommand.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command.mouse

import mono.common.MouseCursor
import mono.graphics.geo.MousePointer
import mono.graphics.geo.PointF
import mono.shape.command.ChangeBound
import mono.shape.shape.AbstractShape
import mono.shapebound.ScaleInteractionPoint
import mono.state.command.CommandEnvironment
import mono.state.utils.UpdateConnectorHelper

/**
 * A [MouseCommand] for scaling shape.
 */
internal class ScaleShapeMouseCommand(
    private val shape: AbstractShape,
    private val interactionPoint: ScaleInteractionPoint
) : MouseCommand {
    override val mouseCursor: MouseCursor? = null

    private val initialBound = shape.bound
    override fun execute(
        environment: CommandEnvironment,
        mousePointer: MousePointer
    ): MouseCommand.CommandResultType {
        when (mousePointer) {
            is MousePointer.Drag ->
                scale(environment, mousePointer.boardCoordinateF, isUpdateConfirmed = false)

            is MousePointer.Up ->
                scale(environment, mousePointer.boardCoordinateF, isUpdateConfirmed = true)

            is MousePointer.Down,
            is MousePointer.Click,
            is MousePointer.DoubleClick,
            is MousePointer.Move,
            MousePointer.Idle -> Unit
        }

        return if (mousePointer == MousePointer.Idle) {
            MouseCommand.CommandResultType.DONE
        } else {
            MouseCommand.CommandResultType.WORKING
        }
    }

    private fun scale(environment: CommandEnvironment, pointF: PointF, isUpdateConfirmed: Boolean) {
        val newBound = interactionPoint.createNewShapeBound(initialBound, pointF)
        environment.shapeManager.execute(ChangeBound(shape, newBound))

        UpdateConnectorHelper.updateConnectors(environment, shape, newBound, isUpdateConfirmed)
        environment.updateInteractionBounds()
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/scale-shape-mouse-command.ts
import type { PointF } from "$libs/graphics-geo/point";
import type { Rect } from "$libs/graphics-geo/rect";
import { ScaleInteractionPoint } from "$mono/shape-interaction-bound/interaction-point";
import { ChangeBound } from "$mono/shape/command/general-shape-commands";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { updateConnectors } from "$mono/state-manager/utils/UpdateShapeBoundHelper";
import { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import { type MousePointer, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";

/**
 * A [MouseCommand] for adding a new Line shape.
 */
export class ScaleShapeMouseCommand implements MouseCommand {
    mouseCursor?: MouseCursor = undefined;
    private readonly initialBound: Rect;

    constructor(private shape: AbstractShape, private interactionPoint: ScaleInteractionPoint.Base) {
        this.initialBound = shape.bound;
    }

    execute(environment: CommandEnvironment, mousePointer: MousePointer): MouseCommand.CommandResultType {
        switch (mousePointer.type) {
            case MousePointerType.DRAG:
                this.scale(environment, mousePointer.boardCoordinateF, false);
                break;
            case MousePointerType.UP:
                this.scale(environment, mousePointer.boardCoordinateF, true);
                break;
            case MousePointerType.DOWN:
            case MousePointerType.MOVE:
            case MousePointerType.CLICK:
            case MousePointerType.DOUBLE_CLICK:
            case MousePointerType.IDLE:
            // nothing to do
        }
        return mousePointer.type === MousePointerType.IDLE ? MouseCommand.CommandResultType.DONE : MouseCommand.CommandResultType.WORKING;
    }

    private scale(environment: CommandEnvironment, pointF: PointF, isUpdateConfirmed: boolean) {
        const newBound = this.interactionPoint.createNewShapeBound(this.initialBound, pointF);
        environment.executeShapeManagerCommand(new ChangeBound(this.shape, newBound));

        updateConnectors(environment, this.shape, newBound, isUpdateConfirmed);
        environment.updateInteractionBounds();
    }
}
